R/utils.R

```diff
@@ -171,6 +172,122 @@ list_init_empty <- function(x,
   list(x = x, sizes = sizes, null = empty_null, typed = empty_typed)
 }
 
+#' Replace `NULL` list elements
+#'
+#' @param x A list, but not a list-of.
+#'
+#' @param sizes An integer vector of sizes of the `x` elements.
+#'
+#' @param ptype One of:
+#'
+#'   - `NULL` to fill `NULL` elements with `unspecified()`.
+#'
+#'   - A `ptype` value to fill `NULL` elements with.
+#'
+#' @param size The size of the replacement value to fill with. Commonly either
+#'   `0L` or `1L`.
+#'
+#' @returns
+#' A named list holding `x` with `NULL`s replaced, and `sizes` updated to
+#' have the correct size where `NULL` elements were replaced.
+#'
+#' @noRd
+list_init_null <- function(x, sizes, ..., ptype = NULL, size = 1L) {
+  check_dots_empty0(...)
+
+  if (!vec_is_list(x)) {
+    cli::cli_abort("`x` must be a list.", .internal = TRUE)
+  }
+  if (is_list_of(x)) {
+    cli::cli_abort("`x` can't be a list-of. Unclass first and provide `ptype`.", .internal = TRUE)
+  }
```

Comment:
The restriction of "no list-of" isn't technically necessary, in particular the previous `list_init_empty()` helper tried to allow list-ofs and use their information. But this is part of what caused the failure in https://github.com/tidyverse/tidyr/issues/1327, and it is cleaner to extract out the `ptype` of list-ofs ahead of time, strip them down to bare lists, and pass all that through instead.

Reply:
This makes sense to me because `list_of()` can't contain `NULL`s, right?

Reply:
They can actually (that's the missing value) but can't contain `unspecified()` (although that branch probably wouldn't be hit for list-ofs because you'd typically also be supplying `ptype`)